src/biconnectivity/articulation.jl

```diff
@@ -22,74 +23,136 @@ julia> articulation(path_graph(5))
 function articulation end
 @traitfn function articulation(g::AG::(!IsDirected)) where {T,AG<:AbstractGraph{T}}
     s = Vector{Tuple{T,T,T}}()
-    is_articulation_pt = falses(nv(g))
     low = zeros(T, nv(g))
     pre = zeros(T, nv(g))
 
+    is_articulation_pt = falses(nv(g))
     @inbounds for u in vertices(g)
-        pre[u] != 0 && continue
-        v = u
-        children = 0
-        wi::T = zero(T)
-        w::T = zero(T)
-        cnt::T = one(T)
-        first_time = true
-
-        # TODO the algorithm currently relies on the assumption that
-        # outneighbors(g, v) is indexable. This assumption might not be true
-        # in general, so in case that outneighbors does not produce a vector
-        # we collect these vertices. This might lead to a large number of
-        # allocations, so we should find a way to handle that case differently,
-        # or require inneighbors, outneighbors and neighbors to always
-        # return indexable collections.
-
-        while !isempty(s) || first_time
-            first_time = false
-            if wi < 1
-                pre[v] = cnt
-                cnt += 1
-                low[v] = pre[v]
-                v_neighbors = collect_if_not_vector(outneighbors(g, v))
-                wi = 1
-            else
-                wi, u, v = pop!(s)
-                v_neighbors = collect_if_not_vector(outneighbors(g, v))
-                w = v_neighbors[wi]
-                low[v] = min(low[v], low[w])
-                if low[w] >= pre[v] && u != v
+        articulation_dfs!(is_articulation_pt, g, u, s, low, pre)
+    end
+
+    articulation_points = T[v for (v, b) in enumerate(is_articulation_pt) if b]
+
+    return articulation_points
+end
+
+"""
+    is_articulation(g, v)
+
+Determine whether `v` is an
+[articulation point](https://en.wikipedia.org/wiki/Biconnected_component) of an undirected
+graph `g`, returning `true` if so and `false` otherwise.
+
+See also [`articulation`].
+
+# Examples
+```jldoctest
+julia> using Graphs
+
+julia> g = path_graph(5)
+{5, 4} undirected simple Int64 graph
+
+julia> articulation(g)
+3-element Vector{Int64}:
+ 2
+ 3
+ 4
+
+julia> is_articulation(g, 2)
+true
+
+julia> is_articulation(g, 1)
+false
+```
+"""
+function is_articulation end
+@traitfn function is_articulation(g::AG::(!IsDirected), v::T) where {T,AG<:AbstractGraph{T}}
+    s = Vector{Tuple{T,T,T}}()
+    low = zeros(T, nv(g))
+    pre = zeros(T, nv(g))
```

Comment:
If the idea is to call `is_articulation` several times in a row for different vertices, these allocations might be reusable? Maybe we should expose `articulation_dfs!`?

Reply:
Sure, that'd be alright with me also. The `_dfs` part is an implementation detail though, so maybe this should just be called `is_articulation!`, taking work-arrays for `pre` and `low` then? Let me know what you think and I'll change to that.

Reply:
Sounds good, but then we should at the very least explain how to initialize these arguments (and ideally what they mean)

Reply:
See also discussion in `https://github.com/JuliaGraphs/Graphs.jl/pull/407#discussion_r1851547909`: maybe we should just be calling these `work_buffer1` and `work_buffer2` (adding some comments to explain in the code - or reassigning them to more aptly named variables) to avoid exposing unnecessary complications and implementation details to users.